components/entities/api/src/main/java/org/phenotips/entities/internal/AbstractPrimaryEntityGroupManager.java

```diff
@@ -17,77 +17,235 @@
- * <ul>
- * <li>the prefix is computed from the uppercase letters of the XClass name, excluding {@code Class}, e.g. for
- * {@code PhenoTips.FamilyGroupClass} the prefix will be {@code FG}; override {@link #getIdPrefix()} to change this
- * behavior</li>
- * <li>the number is a 0-padded 7 digit number, starting at {@code 0000001} and automatically incremented for each new
- * entity created</li>
- * </ul>
- * </li>
- * </ul>
+ * Base class for implementing specific entity groups, where members declare the groups that they belong to. The class
+ * deals with logical group operations and not with implementation.
+ * <p>
  *
- * @param <G> the type of groups handled by this manager
- * @param <E> the type of entities belonging to the groups handled by this manager; if more than one type of entities
- *            can be part of the groups, then a generic {@code PrimaryEntity} should be used instead
+ * @param <G> the type of the group containing the entities
+ * @param <E> the type of entities belonging to this group; if more than one type of entities can be part of the group,
+ *            then a generic {@code PrimaryEntity} should be used instead
  * @version $Id$
  * @since 1.3M2
  */
-@Unstable("New class and interface added in 1.3")
-public abstract class AbstractPrimaryEntityGroupManager<G extends PrimaryEntityGroup<E>, E extends PrimaryEntity>
-    extends AbstractPrimaryEntityManager<G> implements PrimaryEntityGroupManager<G, E>
+public abstract class AbstractPrimaryEntityGroupManager<G extends PrimaryEntity, E extends PrimaryEntity>
+    implements PrimaryEntityGroupManager<G, E>
 {
+    /** Entity type of group object. */
+    protected EntityReference groupEntityReference;
+
+    /** Manager of group primary entities. */
+    protected final PrimaryEntityManager<G> groupManager;
+
+    /** Entity type of member object. */
+    protected EntityReference memberEntityReference;
+
+    /** Manager of member primary entities. */
+    protected final PrimaryEntityManager<E> membersManager;
+
+    /** Logging helper object. */
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @SuppressWarnings("unchecked")
+    protected AbstractPrimaryEntityGroupManager(
+        EntityReference groupEntityReference, EntityReference memberEntityReference)
+    {
+        this.groupEntityReference = groupEntityReference;
+        this.memberEntityReference = memberEntityReference;
+
+        this.membersManager = (PrimaryEntityManager<E>) getManager(memberEntityReference);
+        this.groupManager = (PrimaryEntityManager<G>) getManager(groupEntityReference);
+    }
+
+    private PrimaryEntityManager<PrimaryEntity> getManager(EntityReference entityReference)
+    {
+        PrimaryEntityManager<PrimaryEntity> manager = null;
+        if (memberEntityReference != null) {
+            ComponentManager cm = ComponentManagerRegistry.getContextComponentManager();
```

Comment:
This will fail if `ApplicationStartedEvent` has not been sent yet, because the Component Manager Registry is not initialized yet. 

We learned this because we built a document deletion listener utilizing this method that was being initialized on startup, before `ApplicationStartedEvent` was sent. The application was failing to start and throwing Component Manager exceptions. We eventually narrowed down the problem to this line.

@sdumitriu says that the solution should probably be to use the Component Manager injected directly, as follows, instead of using `ComponentManagerRegistry`.
`@Inject @Named("context") private static Provider<ComponentManager> cmProvider;`

Reply:
This can also be addressed by lazy initialisation of the manager, since it is not needed until something actually happens, which should be after `ApplicationStartedEvent`